### PR TITLE
Feat/browse filters enhancements

### PR DESCRIPTION
**Pull request by Darthmaul**
This PR implements the requirements for the following Trello card: https://trello.com/c/d749LEyt/128-browse-enhance-filters

Implementing some more intricate filtering mechanisms that our users have been wanting for a while. Also implemented a more refined design, as well as a backend scheme which should be easily repeatable moving forward.

Updated form styling: 
image

You can now filter by:

- Loadout Name
- Character (multi-select)
- Primaries (multi-select)
- Secondaries (multi-select)
- Overclocks (y/n)
- Loadout Guide (y/n)
- OC Name specifcially (multi-select)

Implemented using Choicesjs and not any jQuery modules. 

Definitely need a second set of eyes on this backend stuff, and open to any notes on the frontend implementation.

### CODE REVIEW

File: app/Http/Controllers/LoadoutsController.php [hunk 1]
```diff
@@ -14,21 +15,32 @@ class LoadoutsController extends Controller
 {
     public function index(Request $request)
     {
+        // dd($request);
```
Comment (chaseconey) [inline app/Http/Controllers/LoadoutsController.php:18]:
```suggestion
```

File: app/Http/Controllers/LoadoutsController.php [hunk 2]
```diff
@@ -14,21 +15,32 @@ class LoadoutsController extends Controller
...
-
         $loadouts = Loadout::sortable(['updated_at' => 'desc'])
             ->filter($request->all())
             ->with('mods', 'mods.gun', 'character', 'creator')
             ->withCount('votes')
             ->paginate();
         $characters = Character::orderBy('name')->get();
+        $overclocks = Overclock::orderBy('overclock_name')->get();
+        $primaries = Gun::where('character_slot', 1)->orderBy('name')->get();
+        $secondaries = Gun::where('character_slot', 2)->orderBy('name')->get();
```
Comment (chaseconey) [inline app/Http/Controllers/LoadoutsController.php:28]: For these we can probably use `pluck()` to lighten the payload a little bit. That would just create a key value pair for each entry.

File: app/Http/Controllers/LoadoutsController.php [hunk 3]
```diff
@@ -14,21 +15,32 @@ class LoadoutsController extends Controller
...
+            'secondaries' => $secondaries,
+            'overclocks' => $overclocks,
         ]);
     }
 
+    public function getUsersForSearch(Request $request)
+    {
+        return  User::getUsers($request);
+    }
+
```
Comment (chaseconey) [inline app/Http/Controllers/LoadoutsController.php:43]: Is this still used?

File: app/ModelFilters/LoadoutFilter.php [hunk 4]
```diff
@@ -12,15 +12,71 @@ class LoadoutFilter extends ModelFilter
...
+            return $query->whereIn('id', $charIds);
+        });
     }
 
     public function search($name)
     {
         $this->whereLike('loadouts.name', $name);
     }
+
+    public function ocBool($ocVal)
```
Comment (chaseconey) [inline app/ModelFilters/LoadoutFilter.php:32]: Can we change this name to something else. Maybe `hasOverclock()`?

File: app/ModelFilters/LoadoutFilter.php [hunk 5]
```diff
@@ -12,15 +12,71 @@ class LoadoutFilter extends ModelFilter
...
+    }
+
+    public function overclocks($ocIds)
+    {
+        return $this->whereHas('overclocks', function ($query) use ($ocIds) {
+            return $query->whereIn('overclock_id', $ocIds);
+        });
+    }
+
+    public function guide($guideVal)
```
Comment (chaseconey) [inline app/ModelFilters/LoadoutFilter.php:51]: Likewise `hasGuide()`

File: app/User.php [hunk 6]
```diff
@@ -98,4 +99,25 @@ public function loadouts()
     {
         return $this->hasMany(Loadout::class);
     }
+
+    public static function getUsers(Request $request)
```
Comment (chaseconey) [inline app/User.php:103]: We should change the signature here to just pass the search term in so it really doesn't need to know anything about the Request object at all. That will make it easier to test as well :).

File: app/User.php [hunk 7]
```diff
@@ -98,4 +99,25 @@ public function loadouts()
...
+
+        $response = [];
+        foreach ($users as $user) {
+            $response[] = [
+                'id'=>$user->id,
+                'text'=>$user->name,
+            ];
+        }
+
+        return response()->json($response);
```
Comment (chaseconey) [inline app/User.php:121]: We really shouldn't return the response from here either, that should be done in the controller.

File: resources/views/loadouts/index.blade.php [hunk 8]
```diff
@@ -1,51 +1,30 @@
 @extends('layouts.app')
 
+@section('styles')
+<link
+  rel="stylesheet"
+  href="https://cdn.jsdelivr.net/npm/choices.js/public/assets/styles/choices.min.css"
```
Comment (chaseconey) [inline resources/views/loadouts/index.blade.php:6]: Seems like we have `npm install`'ed these (and others) but then we are using a CDN. I vote we remove all the additional package json dependencies and leave this for now. Thoughts?

File: resources/views/loadouts/index.blade.php [hunk 9]
```diff
@@ -1,51 +1,30 @@
...
-                            </svg>
-                        </div>
-                        <input type="search" value="{{ \Request::get('search') }}" name="search" class="focus:ring-gray-500 focus:border-gray-500 block w-full pl-10 sm:text-sm border-gray-300 rounded-md" placeholder="Search Loadouts">
-                    </div>
-                </div>
+
+                      {{-- {{dd(Request::all())}} --}}
+                {{-- @foreach(\Request::all() as $key => $val)
+                    <input type="hidden" name="{{ $key }}" value="{{ is_array($val) ? json_encode($val) : $val }}">
+                @endforeach  --}}
```
Comment (chaseconey) [inline resources/views/loadouts/index.blade.php:25]:
```suggestion
```

**Review by chaseconey — CHANGES_REQUESTED** [review]
Last idea is that we should add a little "Show Filters" button or link and hide most filters (other than search) by default. What do you think about that? It's just a couple of lines of extra work with Alpine and will save lots of space by default.

**Comment by chaseconey** [comment]
I can add some tailwind styles once we merge or towards the end. May make some components out of it as well so we can get easy consistency.

File: app/Http/Controllers/LoadoutsController.php (hunk 2 above)
Comment (Darthmaul) [reply app/Http/Controllers/LoadoutsController.php:28]: Dope

File: app/Http/Controllers/LoadoutsController.php (hunk 3 above)
Comment (Darthmaul) [reply app/Http/Controllers/LoadoutsController.php:43]: This is a remnant of my battle with the search by creator field 😢

File: app/User.php (hunk 7 above)
Comment (Darthmaul) [reply app/User.php:121]: This is a remnant of my battle with the search by creator field 😢

File: resources/views/loadouts/index.blade.php (hunk 8 above)
Comment (Darthmaul) [reply resources/views/loadouts/index.blade.php:6]: I took the `package.json` out back and gave it what for

File: resources/views/loadouts/index.blade.php (hunk 9 above)
Comment (Darthmaul) [reply resources/views/loadouts/index.blade.php:25]: 😅

**Comment by Darthmaul** [comment]
My tasks for today

TODO:

- [ ] Rename "Submit" to "Search"
- [ ] Change "Characters" Placeholder to "Classes"
- [ ] Implement Patch filter
- [ ] Change multi-select dropdown guide to "Click to Select" instead of "Press to Select"
- [ ] Append the name of the class to the dropdown options on Overclocks

File: package.json [hunk 10]
```diff
@@ -47,6 +47,7 @@
         "vue-good-table": "^2.21.10",
         "vue-js-modal": "^2.0.0-rc.6",
         "vue-js-popover": "^1.2.1",
+        "vue-resource": "^1.5.2",
```
Comment (chaseconey) [inline package.json:50]: Do we need this?

**Pull request merged**